### PR TITLE
fix(kad): iterator progress to be decided by any of new peers

### PR DESCRIPTION
**Pull request by maqi**
The `progress` var is against the whole set of incoming closer_peers.
Hence, currently the last new peer will always overwrites the previous result.
i.e. the progress of the iterator is only decided by the last entry of the incoming closer_peers.

This commit is aimed to make the progress to be decided by any of the incoming closer_peer that turns out to be closer to the target.

## Change checklist

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] A changelog entry has been made in the appropriate crates

### CODE REVIEW

**Review by mxinden — COMMENTED** [review]
Well spotted. Out of curiosity, mind sharing how you found the bug?

As far as I can tell, this fortunately only concerns the `QueryStats`.

File: protocols/kad/src/query/peers/closest.rs [hunk 1]
```diff
@@ -175,10 +175,14 @@ impl ClosestPeersIter {
...
-        let num_closest = self.closest_peers.len();
-        let mut progress = false;
-
         // Incorporate the reported closer peers into the iterator.
+        //
+        // The iterator makes progress if:
+        //     1, the iterator did not yet accumulate enough closest peers.
+        //   OR
+        //     2, any of the new peers is closer to the target than any peer seen so far
+        //        (i.e. is the first entry after incorporated)
```
Comment (mxinden) [inline protocols/kad/src/query/peers/closest.rs]:
```suggestion
        //        (i.e. is the first entry after being incorporated)
```

Not a native speaker. I would assume this is easier to understand.

**Comment by mxinden** [comment]
CI failure is unrelated. Seems like cargo semver-checks does not take our `Cargo.lock` into account. It uses `libp2p-yamux` `v0.12.1` which deprecated `WindowUpdateMode`, even thought the corresponding dependabot pull request has not yet been merged. https://github.com/libp2p/rust-libp2p/pull/4930

**Comment by maqi** [comment]
> Well spotted. Out of curiosity, mind sharing how you found the bug?
> 
> As far as I can tell, this fortunately only concerns the `QueryStats`.

We ran a testnet of 2k nodes, and collected the PeerId of each node and the correspondent kBucketsTable from all peers.
According to the collected PeerIds, the avg discovered peers per node is at 150.
Which matchs the mathamatic ideal projection of the nodes' IDs distribution.
However, the real avg discovered peers according to the collected kBucketsTable is just at 124.

We also compared the result of `expected peers in kBucketsTable` and real collected.
The real collected ones are always short of some peers.

Hence I double checked the closest query flow and spot this suspicious point.

File: protocols/kad/src/query/peers/closest.rs (hunk 1 above)
Comment (maqi) [reply protocols/kad/src/query/peers/closest.rs]: PR updated according to the comment

**Comment by thomaseizinger** [comment]
> CI failure is unrelated. Seems like cargo semver-checks does not take our `Cargo.lock` into account. It uses `libp2p-yamux` `v0.12.1` which deprecated `WindowUpdateMode`, even thought the corresponding dependabot pull request has not yet been merged. #4930

cc @obi1kenobi

**Comment by obi1kenobi** [comment]
Thanks for the ping. This is a bit of a tricky situation.

The proximate cause of the semver issue is `RUSTFLAGS: '-Dwarnings'` being set in the job that runs `cargo-semver-checks`: https://github.com/libp2p/rust-libp2p/blob/master/.github/workflows/ci.yml#L15

I don't think it's safe _in general_ for the cargo-semver-checks action to unset `RUSTFLAGS`. I think it's possible that the user may be passing flags other than `-Dwarnings` (e.g. `--cfg <something>`) and removing them may break other use cases.

At the same time, the current behavior is unfortunate and a user-unfriendly default. I have some ideas for this and I'd love your thoughts as well! I've opened https://github.com/obi1kenobi/cargo-semver-checks/issues/589 so we can continue the discussion without derailing this PR further.

> Seems like cargo semver-checks does not take our Cargo.lock into account.

Correct, it does not take `Cargo.lock` into account. This is preferable — the alternative would cause both false-positives and false-negatives.

<details>
<summary>A few reasons: (click to expand)</summary>

- Users of a library experience semver as defined by the crate dependency bounds, not its `Cargo.lock`. If our library re-exports another library's APIs, upgrading that other library to a newer version within the same bounds can cause semver violations we'd like to find. Our users would find them otherwise, since `cargo add` doesn't use the lockfile of libraries.
- Only the "current" version of the library (in cargo-semver-checks terminology) has a lockfile. There's no lockfile for the "baseline" version — the one we're comparing against — so we'd end up using "latest" dependency versions for the baseline. When those dependencies' APIs appear in our own public API, we'll get false positives:
  - the baseline is using a newer version of the dependency which might have made API additions
  - those additions are not present in our library's "current" version, since they aren't in the dependency version in the lockfile
  - therefore the "current" version's public API is smaller than the baseline's public API — this is indistinguishable from a semver-breaking API removal, so it would (falsely) get reported as a breaking change
  - this has happened in the real world: https://github.com/obi1kenobi/cargo-semver-checks/issues/167
</details>

**Comment by obi1kenobi** [comment]
#4942 will avoid this problem for the future in this repo, though I'd still love your feedback on solving the issue within `cargo-semver-checks` itself!

**Comment by thomaseizinger** [comment]
> > Seems like cargo semver-checks does not take our Cargo.lock into account.
> 
> Correct, it does not take `Cargo.lock` into account. This is preferable — the alternative would cause both false-positives and false-negatives.

Thanks for explaining! I guess we just need to update our dependencies ASAP then always :)

**Comment by maqi** [comment]
Hi, @mxinden , 

any idean if this PR is acceptable and to be merged?
thx.

**Review by mxinden — DISMISSED** [review]
Well done!

**Review by mxinden — APPROVED** [review]

**Pull request merged**